q2cli/builtin/tools.py

```diff
@@ -195,7 +196,98 @@ def peek(path):
+              ' parameter is --cast COLUMN:TYPE, repeated for each column'
+              ' and the associated column type it should be cast to in'
+              ' the resultant output.')
+@click.option('--no-error-on-extra', is_flag=True,
+              help='Cast parameters that do not correspond to any of the'
+              ' column names within the provided metadata will result in'
+              ' a raised error unless this flag is enabled.')
+@click.option('--no-ignore-missing', is_flag=True,
+              help='If this flag is enabled, failing to include cast'
+              ' parameters for all columns in the provided metadata will'
+              ' result in an error.')
+@click.option('--output-file', required=True,
+              type=click.Path(exists=False, file_okay=True, dir_okay=False,
+                              writable=True),
+              help='Path to file or directory where the'
+              ' modified metdata should be written to.')
+@click.argument('paths', nargs=-1, required=True, metavar='METADATA...',
+                type=click.Path(exists=True, file_okay=True, dir_okay=False,
+                                readable=True))
+def cast_metadata(paths, cast, output_file, no_error_on_extra,
+                  no_ignore_missing):
+    import tempfile
+    import qiime2
+
+    metadata = _merge_metadata(paths)
+
+    cast_dict = {}
+    try:
+        for casting in cast:
+            k, v = casting.split(':')
+            if k in cast_dict:
+                raise click.BadParameter(
+                    message=(f'Column name "{k}" appears in cast more than'
+                             ' once.'),
+                    param_hint='cast')
+            cast_dict[k] = v
+    except Exception as err:
+        header = \
+            ('Could not parse provided cast arguments into unique key:value'
+             ' pairs. Please make sure all cast flags are of the format --cast'
+             ' COLUMN:TYPE')
+        q2cli.util.exit_with_error(err, header=header)
+
+    valid_types = qiime2.metadata.base.SUPPORTED_COLUMN_TYPES
+    types = set(cast_dict.values())
+    if not types.issubset(valid_types):
+        raise click.BadParameter(
+            message='Invalid column type provided. Please make sure all'
+            ' casted columns include a valid column type.',
+            param_hint='cast')
+
+    column_names = set(metadata.columns.keys())
+    cast_names = set(cast_dict.keys())
+
+    if not no_error_on_extra:
+        if not cast_names.issubset(column_names):
+            raise click.BadParameter(
+                message='One or more cast columns were not found within the'
+                ' metadata.',
+                param_hint='cast')
+
+    if no_ignore_missing:
+        if not column_names.issubset(cast_names):
+            raise click.BadParameter(
+                message='One or more columns within the metadata'
+                ' were not provided in the cast.',
```

Comment:
Same as above, telling the caller what the offending columns are would be really helpful.

Reply:
@thermokarst my only concern with adding this for columns not included in the metadata file is for the scenario that someone only wants to change say, one or two columns in a large metadata file. Do you still think that's reasonable as an output, even if it ends up being a really long list?

Reply:
> the scenario that someone only wants to change say, one or two columns in a large metadata file

If they are doing that, then they wouldn't have the flag enabled, right (which should be the default behavior)? So then this error wouldn't show up for them at all, either way. The point of showing this error is to alert a user who has _opted into_ this behavior. I would show them everything that is wrong, since that's basically what they're asking for.

Reply:
This is a really good question, btw! Thanks for thinking about these things!

Reply:
@thermokarst fair enough, that's a good point - I agree with that! I've updated the logging for this scenario to now include all of the missing columns not included in the cast 🙂